vowpalwabbit/learner.h

```diff
@@ -56,13 +56,19 @@ inline func_data tuple_dbf(void* data, base_learner* base, void (*func)(void*))
 struct learn_data
 {
   using fn = void (*)(void* data, base_learner& base, void* ex);
+  // Label is either label& or std::vector<label>&
+  using fn_with_label = void (*)(void* data, base_learner& base, void* ex, void* label);
   using multi_fn = void (*)(void* data, base_learner& base, void* ex, size_t count, size_t step, polyprediction* pred,
       bool finalize_predictions);
 
   void* data;
   base_learner* base;
   fn learn_f;
   fn predict_f;
+
+  fn_with_label learn_with_label_f;
+  fn_with_label predict_with_label_f;
```

Comment:
This is not making sense to me.  predict should not need label.

Reply:
Eventually yes, but this isn't possible until #2207 is complete

Reply:
Perhaps we should keep this change forward looking.  i.e. if a predict() needs access to label, it can get it from example.  Maybe even introduce some encapsulation in example to get the label.  Eventually label will disappear from example and all reductions will have to conform to this.   

---
In reply to: [388969924](https://github.com/VowpalWabbit/vowpal_wabbit/pull/2323#discussion_r388969924) [](ancestors = 388969924)

Reply:
To have different machinery for predict and learn it pretty difficult especially since most reductions right now use an templated function for both. I see what you're saying but this project is already massive and I really don't want to take on more responsibilities for it. Especially since removing labels from predict is orthogonal to just splitting out the object in the signature.